Reduce precedence of expressions that have an outer attr

dtolnay opened the pull request:
Previously, `-Zunpretty=expanded` would expand this program as follows:

```rust
#![feature(stmt_expr_attributes)]

macro_rules! repro {
    ($e:expr) => {
        #[allow(deprecated)] $e
    };
}

#[derive(Default)]
struct Thing {
    #[deprecated]
    field: i32,
}

fn main() {
    let thing = Thing::default();
    let _ = repro!(thing).field;
}
```

```rs
#![feature(prelude_import)]
#![feature(stmt_expr_attributes)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;

struct Thing {
    #[deprecated]
    field: i32,
}

#[automatically_derived]
impl ::core::default::Default for Thing {
    #[inline]
    fn default() -> Thing {
        Thing { field: ::core::default::Default::default() }
    }
}

fn main() {
    let thing = Thing::default();
    let _ = #[allow(deprecated)] thing.field;
}
```

This is not the correct expansion. The correct output would have `(#[allow(deprecated)] thing).field` with the attribute applying only to `thing`, not to `thing.field`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12457152205/job/34771389231) [(plain)](https://github.com/rust-lang/rust/commit/58242bc6c0b8837d0202e1af4af53c197aef3c28/checks/34771389231/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking parking_lot_core v0.9.10
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1006:56
     |
1006 | ...                   ExprKind::Call(..) => expr.precedence() < ExprPrecedence::Unambiguous || matches!(expr.kind, ExprKind::Field(..)),
     |                                                  ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
1006 |                             ExprKind::Call(..) => expr.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) < ExprPrecedence::Unambiguous || matches!(expr.kind, ExprKind::Field(..)),

error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1007:39
     |
     |
1007 | ...                   _ => expr.precedence() < e.precedence(),
     |                                 ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
1007 |                             _ => expr.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) < e.precedence(),

    Checking annotate-snippets v0.11.5
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1007:56
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1007:56
     |
1007 | ...                   _ => expr.precedence() < e.precedence(),
     |                                                  ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
1007 |                             _ => expr.precedence() < e.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */),

error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1059:35
     |
     |
1059 |                     (prefix, expr.precedence() < ExprPrecedence::Prefix)
     |                                   ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
1059 |                     (prefix, expr.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) < ExprPrecedence::Prefix)

    Checking rustfix v0.8.1
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1161:39
    --> src/tools/clippy/clippy_lints/src/dereference.rs:1161:39
     |
1161 | ...                   if parent.precedence() == ExprPrecedence::Unambiguous {
     |                                 ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
1161 |                             if parent.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) == ExprPrecedence::Unambiguous {

    Checking rustc_version v0.4.1
   Compiling rinja_derive v0.3.5
    Checking regex v1.11.1
    Checking regex v1.11.1
    Checking colored v2.2.0
    Checking levenshtein v1.0.5
    Checking bytes v1.9.0
    Checking pulldown-cmark-escape v0.11.0
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/loops/single_element_loop.rs:87:32
     |
87   |                 arg_expression.precedence() <= ExprPrecedence::Prefix || is_range_literal(arg_expression)
     |                                ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
87   |                 arg_expression.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) <= ExprPrecedence::Prefix || is_range_literal(arg_expression)

    Checking comma v1.0.0
    Checking ui_test v0.26.5
error[E0061]: this method takes 1 argument but 0 arguments were supplied
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/matches/manual_utils.rs:120:75
     |
120  |     let scrutinee_str = if scrutinee.span.eq_ctxt(expr.span) && scrutinee.precedence() < ExprPrecedence::Unambiguous {
     |                                                                           ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
120  |     let scrutinee_str = if scrutinee.span.eq_ctxt(expr.span) && scrutinee.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) < ExprPrecedence::Unambiguous {

    Checking tokio v1.42.0
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/neg_multiply.rs:61:48
    --> src/tools/clippy/clippy_lints/src/neg_multiply.rs:61:48
     |
61   |         let suggestion = if !from_macro && exp.precedence() < ExprPrecedence::Prefix && !has_enclosing_paren(&snip) {
     |                                                ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
61   |         let suggestion = if !from_macro && exp.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) < ExprPrecedence::Prefix && !has_enclosing_paren(&snip) {

error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/redundant_slicing.rs:88:83
     |
     |
88   |             let needs_parens_for_prefix = parent_expr.is_some_and(|parent| parent.precedence() > ExprPrecedence::Prefix);
     |                                                                                   ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
88   |             let needs_parens_for_prefix = parent_expr.is_some_and(|parent| parent.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) > ExprPrecedence::Prefix);

    Checking parking_lot v0.12.3
error[E0061]: this method takes 1 argument but 0 arguments were supplied
    --> src/tools/clippy/clippy_lints/src/transmute/transmutes_expressible_as_ptr_casts.rs:47:19
    --> src/tools/clippy/clippy_lints/src/transmute/transmutes_expressible_as_ptr_casts.rs:47:19
     |
47   |         && parent.precedence() > ExprPrecedence::Cast
     |                   ^^^^^^^^^^-- argument #1 of type `&dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a))` is missing
note: method defined here
    --> /checkout/compiler/rustc_hir/src/hir.rs:1945:12
     |
1945 |     pub fn precedence(
1945 |     pub fn precedence(
     |            ^^^^^^^^^^
help: provide the argument
     |
47   |         && parent.precedence(/* &dyn for<'a> std::ops::Fn(HirId, &'a mut (dyn for<'a> std::ops::FnMut(&'a rustc_hir::Attribute) + 'a)) */) > ExprPrecedence::Cast

    Checking walkdir v2.5.0
    Checking futures v0.3.31
    Checking filetime v0.2.25
```

</details>

fee1-dead commented:
r? compiler

bors commented:
:umbrella: The latest upstream changes (presumably #134788) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dtolnay commented:
Rebased to resolve conflict.

fmease commented:
I'm going to look at this in a few hours.

fmease commented on compiler/rustc_ast/src/ast.rs:
```diff
@@ -1309,11 +1309,20 @@ impl Expr {
+        fn prefix_attrs_precedence(attrs: &AttrVec) -> ExprPrecedence {
+            for attr in attrs {
+                if let AttrStyle::Outer = attr.style {
+                    return ExprPrecedence::Prefix;
```
I think `ExprPrecedence::Prefix` isn't low enough. Consider the following cases involving binops (with `-Zunpretty=expanded`):

```rs
#![feature(stmt_expr_attributes)]

macro_rules! group { ($e:expr) => { $e } }
macro_rules! extra { ($e:expr) => { #[allow()] $e } }

fn main() {                             // master/prefixattr   expected

    let _ = #[allow()] 1 + 1;           // let _ = #[allow()] 1 + 1;
    let _ = group!(#[allow()] 1) + 1;   // let _ = #[allow()] 1 + 1;
    let _ = 1 + group!(#[allow()] 1);   // let _ = 1 + #[allow()] 1;

    let _ = extra!({ 0 }) + 1;          // let _ = #[allow()] { 0 } + 1;
    let _ = extra!({ 0 } + 1);          // let _ = #[allow()] { 0 } + 1;
}
```

I haven't given it much thought, so I don't know which specific precedence level(s) make(s) sense. And indeed, this code is heavily inspired by the discussions in #15701 / #127436. So maybe answering this pretty-printing question is partially blocked on the open design question.

dtolnay replied:
I think that bug is orthogonal to this one.

The bug fixed by this PR is about when an outer expression A contains a subexpression B where B contains an attribute, and in the prettyprinter-produced code the attribute ended up applying to parts of A instead of only to B. This is fixed by having A observe a different effective precedence for its subexpression B, making A generate parentheses around the subexpression containing the attribute, i.e. the attribute will end up **inside** the parentheses.

In contrast, the bug you have identified is about an outer expression A that has an attribute, and a subexpression B that has no attribute. In this case when parentheses are inserted, the attribute will end up **outside** the parentheses.

I will fix that as well but I expect it will involve unrelated code.

dtolnay replied:
Fix: https://github.com/rust-lang/rust/pull/142476

bors commented:
:umbrella: The latest upstream changes (presumably #137235) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dtolnay commented:
Rebased to resolve conflicts with:

- https://github.com/rust-lang/rust/pull/137504
- https://github.com/rust-lang/rust/pull/140450
- https://github.com/rust-lang/rust/pull/142240

dtolnay committed:
Add AST pretty-printer tests involving attr precedence

This test currently fails (as expected).

    --- stderr -------------------------------
    Pretty-printer lost necessary parentheses
      BEFORE: (#[attr] loop {}).field
       AFTER: #[attr] loop {}.field
    ------------------------------------------

dtolnay committed:
Reduce precedence of expressions that have an outer attr

dtolnay commented:
- Rebased to resolve conflict with https://github.com/rust-lang/rust/pull/140770

fmease commented on tests/ui-fulldeps/auxiliary/parser.rs:
```diff
@@ -46,4 +48,36 @@ impl MutVisitor for Normalize {
     fn visit_span(&mut self, span: &mut Span) {
         *span = DUMMY_SP;
     }
```
Could you remind we why we need normalization in the first place? The downstream crates seem to only compare exprs by their pretty-printed representation (granted, AST nodes don't impl Eq) and AST pretty only cares about spans for emitting comments which these test cases don't contain.

Due to that knowledge gap I'm surprised that you can get away with not erasing *all* spans in `visit_attribute` (e.g., the ones contained in `normal_attr.item`)

dtolnay replied:
This matters for the `format!("{expr:#?}") != format!("{expr2:#?}")` in this part of the test:

https://github.com/rust-lang/rust/blob/8da623945f83933dd38644d5745532ee032e855b/tests/ui-fulldeps/pprust-parenthesis-insertion.rs#L182-L193

Without erasing the Spans and NodeIds and AttrIds, the comparison fails spuriously depending on whitespace.

```diff
Attribute {
    kind: Normal(
        NormalAttr {
            ...
            tokens: Some(
-               LazyAttrTokenStream(AttrTokenStream([Token(Token { kind: Pound, span: Span { lo: BytePos(528), hi: BytePos(529), ctxt: #0 } }, Joint), Token(Token { kind: Bang, span: Span { lo: BytePos(529), hi: BytePos(530), ctxt: #0 } }, JointHidden), Delimited(DelimSpan { open: Span { lo: BytePos(530), hi: BytePos(531), ctxt: #0 }, close: Span { lo: BytePos(535), hi: BytePos(536), ctxt: #0 } }, DelimSpacing { open: JointHidden, close: Alone }, Bracket, AttrTokenStream([Token(Token { kind: Ident("attr", No), span: Span { lo: BytePos(531), hi: BytePos(535), ctxt: #0 } }, JointHidden)]))])),
+               LazyAttrTokenStream(AttrTokenStream([Token(Token { kind: Pound, span: Span { lo: BytePos(560), hi: BytePos(561), ctxt: #0 } }, Joint), Token(Token { kind: Bang, span: Span { lo: BytePos(561), hi: BytePos(562), ctxt: #0 } }, JointHidden), Delimited(DelimSpan { open: Span { lo: BytePos(562), hi: BytePos(563), ctxt: #0 }, close: Span { lo: BytePos(567), hi: BytePos(568), ctxt: #0 } }, DelimSpacing { open: JointHidden, close: Alone }, Bracket, AttrTokenStream([Token(Token { kind: Ident("attr", No), span: Span { lo: BytePos(563), hi: BytePos(567), ctxt: #0 } }, JointHidden)]))])),
            ),
```

There are 2 other possible ways to express this `assert(expr == parse(print(expr)))` test:

- Continue using Debug (`{:#?}`), but instead of performing `Unparenthesize.visit_expr(&mut expr)` beforehand, use a regex to normalize the rendered representation, i.e. equivalent to `sed 's/(BytePos|AttrId|NodeId)\([0-9]+\)//'`

- Do not use Debug, use some other PartialEq-like trait to traverse two Exprs and perform the comparison ignoring spans. Syn's test suite actually has an implementation of this (for rustc's data structures! not syn's data structures) but it is 900 lines. https://github.com/dtolnay/syn/blob/2.0.103/tests/common/eq.rs

As for why `normal_attr.item` does not need to be handled by the code added by this PR, that is because it's handled by the `mut_visit::walk_attribute(self, attr)`.

https://github.com/rust-lang/rust/blob/8da623945f83933dd38644d5745532ee032e855b/compiler/rustc_ast/src/visit.rs#L1630-L1654

The code here in the test only needs to deal with the attribute tokens, which are not visited by `mut_visit` since https://github.com/rust-lang/rust/pull/140450. Prior to https://github.com/rust-lang/rust/pull/140450, this code was unnecessary and we could just use `const VISIT_TOKENS: bool = true`.

fmease commented:
Some comments incl. questions but once addressed in one way or another, r=me

dtolnay commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -2285,12 +2285,23 @@ pub struct Expr<'hir> {
-    pub fn precedence(&self) -> ExprPrecedence {
+    pub fn precedence(
+        &self,
+        for_each_attr: &dyn Fn(HirId, &mut dyn FnMut(&Attribute)),
```
The `HirId` argument is here because `for_each_attr` needs to be able to iterate not only `self`'s attributes, but potentially some other Expr's attributes in the `ExprKind::DropTemps` case.

If the `HirId` were dropped from `for_each_attr`, the callers would need to change like this:

```diff
-    fn precedence(&self, expr: &hir::Expr<'_>) -> ExprPrecedence {
-        let for_each_attr = |id: HirId, callback: &mut dyn FnMut(&hir::Attribute)| {
-            self.attrs(id).iter().for_each(callback);
+    fn precedence(&self, mut expr: &hir::Expr<'_>) -> ExprPrecedence {
+        while let hir::ExprKind::DropTemps(inner, ..) = expr.kind {
+            expr = inner;
+        }
+        let for_each_attr = |callback: &mut dyn FnMut(&hir::Attribute)| {
+            self.attrs(expr.hir_id).iter().for_each(callback);
         };
         expr.precedence(&for_each_attr)
     }
```

which is fine but I think less straightforward than the current approach.

dtolnay commented:
@bors r=fmease

dtolnay commented:
I think that is misattributed. That exact same failure also appears in https://github.com/rust-lang/rust/pull/142533 which does not include this PR.

@bors r=fmease

bors commented:
:pushpin: Commit 2171f89eb223a4f5f3af74df24104d77c1a8a6aa has been approved by `fmease`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fmease 2171f89eb223a4f5f3af74df24104d77c1a8a6aa -->
<!-- homu: {"type":"Approved","sha":"2171f89eb223a4f5f3af74df24104d77c1a8a6aa","approver":"fmease","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
